File: sim/state/lcd.ts
/// <reference path="../../node_modules/pxt-core/built/pxtsim.d.ts"/>
/// <reference path="../../node_modules/pxt-core/localtypings/pxtarget.d.ts"/>
/// <reference path="../../built/common-sim.d.ts"/>
namespace pxsim.LCD {
    export function setCursor(x: number, y: number) {
        let b = lcdState();
        if (y < b.lines && y >= 0)
            b.cursor[0] = y;
        if (x < b.columns && x >= 0)
            b.cursor[1] = x;
        else if (x >= b.columns)
            b.cursor[1] = b.columns;  
        b.setUsed();
    }

    export function ShowNumber(n: number) {
        ShowString("" + n);
    }

    export function ShowValue(name: string, value: number) {
        ShowString( name + ":" + value);
    }

    export function ShowString(s: string) {
        let b = lcdState();
        b.setUsed();

        if (b.cursor[0] >= b.lines || b.cursor[0] < 0)
            return;
            
        if (b.cursor[1] >= b.columns || b.cursor[1] < 0)
            return;

        b.text[b.cursor[0]] = b.text[b.cursor[0]].substring(0, b.cursor[1]) + s + b.text[b.cursor[0]].substring(b.cursor[1] + s.length, b.columns);
        b.cursor[1] += s.length;
        runtime.queueDisplayUpdate()
    }

    export function clear() {
        let b = lcdState();
        b.text[0] = "                ";
        b.text[1] = "                ";
        b.setUsed();
    }

    export function SetBacklightColor(rgb: number) {
        let b = lcdState();
        b.backLightColor = RGBColorToHtmlColor(rgb);
        b.setUsed();
    }

    export function BacklightOn() {
        let b = lcdState();
        b.backLightColor = "#A0F7F7";
        b.setUsed();

    }

    export function BacklightOff() {
        let b = lcdState();
        b.backLightColor = "#6e7d6e";
        b.setUsed();
    }

    export function ShowNumberAtPos(n: number, x: number, y: number) {
        setCursor(x, y);
        ShowNumber(n);
        runtime.queueDisplayUpdate()
    }

    export function ShowStringAtPos(s: string, x: number, y: number) {
        setCursor(x, y);
        ShowString(s);
        runtime.queueDisplayUpdate()

    }

    function RGBColorToHtmlColor(rgb: number): string {
        let red = unpackR(rgb);
        let green = unpackG(rgb);
        let blue = unpackB(rgb);
        let html = "#" +
            (red > 10 ? red.toString(16) : "0" + red.toString(16)) +
            (green > 10 ? green.toString(16) : "0" + green.toString(16)) +
            (blue > 10 ? blue.toString(16) : "0" + blue.toString(16));
        return html;
    }

    function unpackR(rgb: number): number {
        let r = (rgb >> 16) & 0xFF;
        return r;
    }
    function unpackG(rgb: number): number {
        let g = (rgb >> 8) & 0xFF;
        return g;
    }
    function unpackB(rgb: number): number {
        let b = (rgb >> 0) & 0xFF;
        return b;
    }

}

File: libs/lcd/LCD.cpp
#include "pxt.h"
#include "target_lcd.h"

namespace pxt {
SINGLETON(WLCD);
}

namespace LCD {
    uint8_t unpackR(uint32_t rgb) {
        uint8_t r = (rgb >> 16) & 0xFF;
        return r;
    }
    uint8_t unpackG(uint32_t rgb) {
        uint8_t g = (rgb >> 8) & 0xFF;
        return g;
    }
    uint8_t unpackB(uint32_t rgb) {
        uint8_t b = (rgb >> 0) & 0xFF;
        return b;
    }


    /**
     * set cursor position at given position
     * @param x is LCD column position, eg: 0
     * @param y is LCD row position, eg: 0
     */
    //% blockId="LCD_SET_CURSOR" 
    //% block="set cursor position |at x %x|y %y"
    //% weight=89 
    //% blockGap=8 
    //% x.min=0 x.max=15
    //% y.min=0 y.max=1
    //% parts=LCD trackArgs=0
    void setCursor(uint8_t x ,uint8_t y){
        getWLCD()->lcd.setCursor(x, y);
    }

    /**
     * show a number in LCD at given position
     * @param n is number will be show, eg: 10, 100, 200
     */
    //% blockId="LCD_SHOW_NUMBER" 
    //% block="show number %n"
    //% weight=90 
    //% blockGap=8
    //% parts=LCD trackArgs=0
    void ShowNumber(uint32_t n){
        getWLCD()->lcd.print(n);
    }

    /**
     * show a string in LCD at given position
     * @param s is string will be show, eg: "Hello"
     */
    //% blockId="LCD_SHOW_STRING" 
    //% block="show string %s"
    //% weight=90 
    //% blockGap=8
    //% parts=LCD trackArgs=0
    void ShowString(String s) {
        getWLCD()->lcd.print(s->data);
    }

    /**
     * show a name:value pair in LCD at given position
     * @param name : a string that is the name part of the name:value pair, eg: "x"
     * @param value : a number that is the value part of the name:value pair, eg: 0
     */
    //% blockId="LCD_SHOW_VALUE" 
    //% block="show value %name = %value"
    //% weight=90 
    //% blockGap=8
    //% parts=LCD trackArgs=0
    void ShowValue(String name, uint32_t value) {
        getWLCD()->lcd.print(name->data);
        getWLCD()->lcd.print(':');
        getWLCD()->lcd.print(value);
    }

    /**
     * clear all display content
     */
    //% blockId="LCD_CLEAR" 
    //% block="clear screen"
    //% weight=85 
    //% blockGap=8
    //% parts=LCD trackArgs=0
    void clear(){
        getWLCD()->lcd.clear();
    }

    /**
     * set LCD backlight color
     * 
     * @param rgb RGB color of the backlight LED
     */
    //% blockId="LCD_BACKLIGHT_COLOR" 
    //% block="set backlight color %rgb=colorNumberPicker"
    //% weight=87
    //% blockGap=8
    //% parts=LCD trackArgs=0
    void SetBacklightColor(uint32_t rgb){
        uint8_t r = unpackR(rgb);
        uint8_t g = unpackG(rgb);
        uint8_t b = unpackB(rgb);
        getWLCD()->lcd.setRGB(r, g, b);
    }

    /**
     * turn on LCD backlight
     */
    //% blockId="LCD_BACKLIGHT_ON" 
    //% block="turn on backlight"
    //% group="More" 
    //% weight=71 
    //% blockGap=8
    //% parts=LCD trackArgs=0
    void BacklightOn(){
        getWLCD()->lcd.setColorWhite();
    }

    /**
     * turn off LCD backlight
     */
    //% blockId="LCD_BACKLIGHT_OFF" 
    //% block="turn off backlight"
    //% group="More" 
    //% weight=70 
    //% blockGap=8
    //% parts=LCD trackArgs=0
    void BacklightOff(){
        getWLCD()->lcd.setColorAll();

    }

    /**
     * show a number in LCD at given position
     * @param n is number will be show, eg: 10, 100, 200
     * @param x is LCD column position, eg: 0
     * @param y is LCD row position, eg: 0
     */
    //% blockId="LCD_SHOW_NUMBER_AT_POS" 
    //% block="show number %n|at x %x|y %y"
    //% group="More" 
    //% weight=10 
    //% blockGap=8
    //% x.min=0 x.max=15
    //% y.min=0 y.max=1
    //% parts=LCD trackArgs=0
    void ShowNumberAtPos(uint32_t n ,uint8_t x ,uint8_t y){
        getWLCD()->lcd.setCursor(x, y);
        ShowNumber(n);
    }

    /**
     * show a string in LCD at given position
     * @param s is string will be show, eg: "Hello"
     * @param x is LCD column position, [0 - 15], eg: 0
     * @param y is LCD row position, [0 - 1], eg: 0
     */
    //% blockId="LCD_SHOW_STRING_AT_POS" 
    //% block="show string %s|at x %x|y %y"
    //% group="More" 
    //% weight=10 
    //% blockGap=8
    //% x.min=0 x.max=15
    //% y.min=0 y.max=1
    //% parts=LCD trackArgs=0
    void ShowStringAtPos(String s, uint8_t x, uint8_t y) {
        getWLCD()->lcd.setCursor(x, y);
        ShowString(s);
    }
}